### DISCUSSION TITLE
Renovate not working with internal hosts

### DISCUSSION THREAD

**Discussion by rmr2014** [renovatebot/renovate#11501] [Suggest an Idea]
Hello,

No matter what how many times I create config.js file with my 'matchHost' or 'registryUrl' entry for my internal artifactory, it never works. It always looks for outside like 'maven registry' or 'docker.io' etc, please advise on this

error:
Failed to look up dependency

**Comment by rarkins** [comment]
`hostRules` control which authentication is used. It sounds like you need to define `packageRules` with `registryUrls` to control which registry host is used.

**Reply by rmr2014** [reply]
Little background:
I created a config.js with base config and all other details and also added both package rules and registry url for docker. Since I am using harbor to get docker details using that url. 

problem is, my config file able to fetch dependencies but it never able to create a merge request or it always fails to lookup dependency.

Could you please help me for a simple dockerfile dependency update.

**Reply by rarkins** [reply]
Unfortunately, you are not providing enough information to be helped

**Reply by rmr2014** [reply]
I am using Gitlab tool and using hostrules to autenticate my harbor tool where I have my docker registry with alpine image.

I created a dockerfile with alpine:3.1 version and now I wanna use my renovate config to check the version and create a merge request for higher version.

My config file able to fetch dependency as 3.1 from my dockerfile but its unable to lookup the higher version of alpine from my harbor tool

**Reply by viceice** [reply]
Please provide your sanitized config and a small sample file, which should be updated

**Comment by rmr2014** [comment]
This is my error: it always looking at "https://repo.maven.apache.org/maven2" though I have provided hostrules and packagerules.

Error:
`"Failed to look up dependency junit:junit"}],"versioning":"maven"},{"datasource":"maven","depName":"org.apache.maven.plugins:maven-compiler-plugin","currentValue":"3.1","fileReplacePosition":1326,"registryUrls":["https://repo.maven.apache.org/maven`

My config.js file:

```
> module.exports = {
>   platform: 'gitlab',
>   endpoint: '<on-primise gitlab>',
>   assignees: ['<my id>'],
>   gitAuthor: "<my email>",
>   baseBranches: ['master'],
>   labels: ['renovate'],
>   automerge: false,
>   automergeType: 'branch',
>   logFileLevel: "debug",
>   logFile: "renovate.log",
>   docker: {"enabled": false },
> 
>   repositories: [
>     "<my repo name>"
>   ],
> 
>   maven: {
> 	    "enabled": true
> 	},
> 	hostRules: [{
> 	    "platform": "maven",
>             "endpoint": "https://<my domain.com>/maven-repo/",
> 	    "username": "process.env.USER",    
> 	    "password": "process.env.PASSWORD"
> 	},],
> 	
>      "packageRules": [{
>      "managers": ["maven"],
> 	   "registryUrls": ["https://<my domain.com>/maven-repo/"]
>     }],
> 	onboardingConfig: { extends: ["config:base"] }
> };
```

I am using simple pom.xml file to check the dependencies and using renovate image: renovate:25.72

**Reply by viceice** [reply]
Host and package rules should be top level config keys.

Please format your config as JSON code.

**Comment by rmr2014** [comment]
I got merge requests but I didn't understand one thing, why it is looking at github.com as source for my plugins. Please check below details.

 ```

> WARN: No github.com token has been configured. Skipping release notes retrieval (repository=myRepo)
>        "manager": "maven",
>        "depName": "junit:junit",
>        "sourceUrl": "https://github.com/junit-team/junit4"
>  WARN: No github.com token has been configured. Skipping release notes retrieval (repository=myRepo)
>        "manager": "maven",
>        "depName": "junit:junit",
>        "sourceUrl": "https://github.com/junit-team/junit4"

```

**Reply by rarkins** [reply]
Because `https://github.com/junit-team/junit4` has been identified as the source of `junit`

**Reply by rmr2014** [reply]
Please correct me if my understanding is correct:
1. It checks provided plugin versions/dependencies in my pom.xml and extract them.
2. Then it will check if any higher version available in my own repository which is given under hostRules
3. Then if it find any higher one then it creates a merge request to upgrade the dependency.

**Comment by rmr2014** [comment]
Hi,

I am able to configure my maven project but for javascript I am getting this below error:

`INFO: External host error causing abort - skipping`

My config file

```
module.exports = {
>   platform: 'gitlab',
>   endpoint: '<on-primise gitlab>',
>   assignees: ['<my id>'],
>   gitAuthor: "<my email>",
>   baseBranches: ['master'],
>   labels: ['renovate'],
>   automerge: false,
>   automergeType: 'branch',
>   logFileLevel: "debug",
>   logFile: "renovate.log",
>   docker: {"enabled": false },
> 
>   repositories: [
>     "<my repo name>"
>   ],
> 
> 	hostRules: [{
> 	    "platform": "npm",
>             "endpoint": "https://<my domain.com>/maven-repo/",
> 	    "username": "process.env.USER",    
> 	    "password": "process.env.PASSWORD"
> 	},],
> 	
>      "packageRules": [{
>      "managers": ["npm"],
> 	   "registryUrls": ["https://<my domain.com>/npm-public/"]
>     }],
> 	onboardingConfig: { extends: ["config:base"] }
> };
```

Also in logs I see below socketError:

`"stack": "RequestError: Timeout awaiting 'request' for 60000ms\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:956:65)\n    at Object.onceWrapper (events.js:483:26)\n    at ClientRequest.emit (events.js:388:22)\n    at ClientRequest.emit (domain.js:470:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:376:20)\n    at TLSSocket.emit (domain.js:470:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at Timeout.timeoutHandler [as _onTimeout] (/usr/src/app/node_modules/got/dist/source/core/utils/timed-out.js:36:25)\n    at listOnTimeout (internal/timers.js:557:17)\n    at processTimers (internal/timers.js:498:7)",`

**Reply by viceice** [reply]
Check the docs, some settings are outdated.

Your host rule and package rule use different urls?

**Reply by rmr2014** [reply]
No, both are same
hostRules: [{
> 	    "platform": "npm",
>             "endpoint": "https://<my domain.com>/maven-repo/",
> 	    "username": "process.env.USER",    
> 	    "password": "process.env.PASSWORD"
> 	},],
> 	
>      "packageRules": [{
>      "managers": ["npm"],
> 	   "registryUrls": ["https://<my domain.com>/npm-public/"]
>     }],

**Reply by viceice** [reply]
for javascript you need to add `registry=https://<my domain.com>/npm-public/` to your repo `.npmrc`. 

Overriding `registryUrls` is not yet supported for `npm` manager.

**Reply by rmr2014** [reply]
Yes I already added that and removed registryUrl in my config.js file but still getting the below error
**INFO: External host error causing abort - skipping (repository=<my repo name>)**


Config.js file is below

```
module.exports = {
>   platform: 'gitlab',
>   endpoint: '<on-primise gitlab>',
>   assignees: ['<my id>'],
>   gitAuthor: "<my email>",
>   baseBranches: ['master'],
>   labels: ['renovate'],
>   automerge: false,
>   automergeType: 'branch',
>   logFileLevel: "debug",
>   logFile: "renovate.log",
>   docker: {"enabled": false },
> 
>   repositories: [
>     "<my repo name>"
>   ],
> 
> 	hostRules: [{
> 	
      hostType: 'npm',
      matchHost: "<myDomain/npm-public/>",
>    username: process.env.USER,    
>    password: process.env.PASSWORD
> 	},],
> 	
> 	onboardingConfig: { extends: ["config:base"] }
> };
```

**Reply by viceice** [reply]
please format the config as code

**Reply by rmr2014** [reply]
```js
module.exports =
{
  platform: 'gitlab',
  endpoint: '<on-primise gitlab',
  assignees: ['<my id'],
  gitAuthor: "<my email",
  baseBranches: ['master'],
  labels: ['renovate'],
  automerge: false,
  automergeType: 'branch',
  logFileLevel: "debug",
  logFile: "renovate.log",
  docker:
  {
    "enabled": false
  },

  repositories: [
    "<my repo name>"
  ],

  hostRules: [
    {

      hostType: 'npm',
      matchHost: "<myDomain.com/npm-public/>",
      username: process.env.USER,
      password: process.env.PASSWORD
    },
  ],

  onboardingConfig:
  {
    extends: ["config:base"]
  }
};

```

**Reply by viceice** [reply]
@rmr2014 What's the difficulty to format the code like above?

please provide the debug logs (formatted)

**Reply by rmr2014** [reply]
Okay I will share the log. 

I tried formatting multiple times here in edit box using that code format (using this <>), but it never worked for me not sure why, thats the reason its not properly formatted.

**Reply by viceice** [reply]
https://guides.github.com/features/mastering-markdown/

**Reply by rmr2014** [reply]
I am getting 401 unauthorized error, below are the logs. Though I am authenticating properly but still seeing this error

`{"name":"renovate","hostname":"runner-jwghfsh7-project-1710-concurrent-0rv9bw","pid":32,"level":20,"logContext":"DcSmUvDEl","repository":"<my reponame>","packageUrl":"<myDomain.com/npm-public>/ts-node","err":{"name":"HTTPError","timings":{"start":1632121886206,"socket":1632121886206,"lookup":1632121886256,"connect":1632121886267,"secureConnect":1632121886271,"upload":1632121886271,"response":1632121886286,"end":1632121886287,"phases":{"wait":0,"dns":50,"tcp":11,"tls":4,"request":0,"firstByte":15,"download":1,"total":81}},"message":"Response code 401 (Unauthorized)","stack":"HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)","options":{"headers":{"user-agent":"RenovateBot/25.72.3 (https://github.com/renovatebot/renovate)","authorization":"***********","accept":"application/json","accept-encoding":"gzip, deflate, br"},"url":"<myDomain.com/npm-public>/ts-node","username":"","password":"","method":"GET","http2":false},"response":{"statusCode":401,"statusMessage":"Unauthorized","body":"","headers":{"server":"nginx/1.13.7","date":"Mon, 20 Sep 2021 07:11:26 GMT","transfer-encoding":"chunked","connection":"close","x-content-type-options":"nosniff","www-authenticate":"BASIC realm=\"Sonatype Nexus Repository Manager\""},"httpVersion":"1.1"}},"statusCode":401,"packageName":"ts-node","msg":"Dependency lookup failure: unauthorized","time":"2021-09-20T07:11:26.287Z","v":0}`

**Reply by viceice** [reply]
- make sure the `matchHost` from hostrule is equal to the `registry=` line in your npmrc and endswith a slash.
- make sure you don't have any `RENOVATE_PASSWORD` variables beside `RENOVATE_TOKEN` in your env, otherwise they can override your host rules.
 
Check for these log messages:

https://github.com/renovatebot/renovate/blob/8fb9197d0cbb35aece07597f694d3bf851261fe0/lib/util/host-rules.ts#L33-L35

https://github.com/renovatebot/renovate/blob/7b3e14c5cc741457b2ab1f98a7bd9bb90bd01d00/lib/workers/global/config/parse/env.ts#L123-L123

**Reply by rmr2014** [reply]
Yes all looks fine and this time error is different

`{"name":"renovate","hostname":"runner-hdawdoms-project-1710-concurrent-0lxxq2","pid":33,"level":40,"logContext":"-MDH_1rag","repository":"EMEA/SCM/TESTING/ravi/Javascript-template","hostType":"npm","lookupName":"js-yaml","err":{"name":"TimeoutError","code":"ETIMEDOUT","timings":{"start":1632234098346,"socket":1632234098347,"lookup":1632234098371,"error":1632234158398,"phases":{"wait":1,"dns":24,"total":60052}},"event":"request","message":"Timeout awaiting 'request' for 60000ms","stack":"RequestError: Timeout awaiting 'request' for 60000ms\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:956:65)\n    at Object.onceWrapper (events.js:483:26)\n    at ClientRequest.emit (events.js:388:22)\n    at ClientRequest.emit (domain.js:470:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:376:20)\n    at TLSSocket.emit (domain.js:470:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at Timeout.timeoutHandler [as _onTimeout] (/usr/src/app/node_modules/got/dist/source/core/utils/timed-out.js:36:25)\n    at listOnTimeout (internal/timers.js:557:17)\n    at processTimers (internal/timers.js:498:7)","options":{"headers":{"user-agent":"RenovateBot/25.72.3 (https://github.com/renovatebot/renovate)","accept":"application/json","accept-encoding":"gzip, deflate, br"},"url":"https://registry.npmjs.org/js-yaml","username":"","password":"","method":"GET","http2":false}},"msg":"Host error","time":"2021-09-21T14:22:38.400Z","v":0}`

Also, 34 line is showing in my renovate logs

`{"name":"renovate","hostname":"runner-hdawdoms-project-1710-concurrent-0lxxq2","pid":33,"level":20,"logContext":"-MDH_1rag","msg":"Adding token authentication for mygitlab.com to hostRules","time":"2021-09-21T14:19:20.057Z","v":0}`

but 123 line I can't see in logs

**Reply by viceice** [reply]
you must see `Adding password authentication for  <myDomain.com/npm-public/> to hostRules`. if not, you've made something wrong.

How are you running renovate? It seems a self-hosted GitLab?

We've official templates for running gitlab pipelines at https://gitlab.com/renovate-bot/renovate-runner

**Reply by rmr2014** [reply]
I am using image renovate:25.72 and similar configuration working for my maven pom.xml, for npm only not working.
also I don't see "`Adding password authentication for <myDomain.com/maven-public/> to hostRules`" in that maven build

**Reply by viceice** [reply]
you need to add `always-auth=true` to your `.npmrc`

```ini
registry=httpsa://myDomain.com/npm-public/
//myDomain.com/npm-public/:always-auth=true
```

ref https://github.com/renovatebot/renovate/discussions/11284#discussioncomment-1368867

**Reply by rmr2014** [reply]
Hi, 
Added this parameter but no change, basically my internal host has public npm repo which doesn't need authentication, I am seeing this below error everytime, looks like its unable to get dependencies from source (github) and then failed to look higher version of dependencies from my nexus repo.

> INFO: Repository started (repository=my repoName)
       "renovateVersion": "25.72.3"
 INFO: Dependency extraction complete (repository=my repoName)
       "baseBranch": "master",
       "stats": {
         "managers": {"npm": {"fileCount": 2, "depCount": 31}},
         "total": {"fileCount": 2, "depCount": 31}
       }
 WARN: Host error (repository=my repoName)
       "hostType": "npm",
       "lookupName": "js-yaml",
       "err": {
         "name": "TimeoutError",
         "code": "ETIMEDOUT",
         "timings": {
           "start": 1632474162939,
           "socket": 1632474162939,
           "lookup": 1632474163045,
           "error": 1632474222998,
           "phases": {"wait": 0, "dns": 106, "total": 60059}
         },
         "event": "request",
         "message": "Timeout awaiting 'request' for 60000ms",
          "stack": "RequestError: Timeout awaiting 'request' for 60000ms\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:956:65)\n    at Object.onceWrapper (events.js:483:26)\n    at ClientRequest.emit (events.js:388:22)\n    at ClientRequest.emit (domain.js:470:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:376:20)\n    at TLSSocket.emit (domain.js:470:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at Timeout.timeoutHandler [as _onTimeout] (/usr/src/app/node_modules/got/dist/source/core/utils/timed-out.js:36:25)\n    at listOnTimeout (internal/timers.js:557:17)\n    at processTimers (internal/timers.js:498:7)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/25.72.3 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://registry.npmjs.org/js-yaml",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         }
       }
 INFO: External host error causing abort - skipping (repository=my repoName)
 INFO: Repository finished (repository=my repoName)
       "durationMs": 205517
>

**Reply by viceice** [reply]
your npmrc config must be wrong, because renovate is using default npmjs registry. 

make sure you have a trailing newline in `.npmrc` otherwise you'll get issues.

**Reply by rmr2014** [reply]
OK, this is my .npmrc file

`loglevel=info`
`registry=https://mydomain.com/repository/npm-public/`
`strict-ssl=false
`

**Reply by rmr2014** [reply]
Also I am seeing this error in log

{"name":"renovate","hostname":"runner-qvfei-ya-project-1710-concurrent-0bhslr","pid":37,"level":20,"logContext":"A6diI7Hqo","repository":"<my repoName>","dependency":"@types/node","packageFile":"package.json","msg":"Failed to look up dependency @types/node","time":"2021-09-24T09:40:12.211Z","v":0}

**Reply by rmr2014** [reply]
Hi , thanks for ur support I have fixed the issue.
Simple thing added ":disableHost(registry.npmjs.org/js-yaml)" line in my config.js which disable this host "https://registry.npmjs.org/js-yaml",

` extends: [
    'config:base',
    ":onlyNpm",
    ":pinAllExceptPeerDependencies",
    ":disableHost(registry.npmjs.org/js-yaml)"

  ],

`
